user-org-keycloak-util/src/main/java/org/sunbird/sso/service/impl/KeyCloakServiceImpl.java

```diff
@@ -0,0 +1,486 @@
+package org.sunbird.sso.service.impl;
+
+import static java.util.Arrays.asList;
+
+import java.security.KeyFactory;
+import java.security.PublicKey;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.lang3.StringUtils;
+import org.keycloak.RSATokenVerifier;
+import org.keycloak.admin.client.Keycloak;
+import org.keycloak.admin.client.resource.UserResource;
+import org.keycloak.representations.AccessToken;
+import org.keycloak.representations.idm.CredentialRepresentation;
+import org.keycloak.representations.idm.UserRepresentation;
+import org.sunbird.exception.AuthorizationException;
+import org.sunbird.exception.ProjectCommonException;
+import org.sunbird.exception.message.IResponseMessage;
+import org.sunbird.exception.message.Localizer;
+import org.sunbird.exception.message.ResponseCode;
+import org.sunbird.sso.model.User;
+import org.sunbird.sso.service.ISSOService;
+import org.sunbird.sso.util.KeyCloakConnectionProvider;
+import org.sunbird.sso.util.SSOConstant;
+import org.sunbird.util.LoggerEnum;
+import org.sunbird.util.ProjectLogger;
+import org.sunbird.util.PropertiesCache;
+
+/**
+ * Single sign out service implementation with Key Cloak.
+ *
+ * @author Manzarul
+ */
+public class KeyCloakServiceImpl implements ISSOService {
+
+  private Keycloak keycloak = KeyCloakConnectionProvider.getConnection();
+  private static PublicKey SSO_PUBLIC_KEY = null;
+
+  public PublicKey getPublicKey() {
+    if (null == SSO_PUBLIC_KEY) {
+      SSO_PUBLIC_KEY =
+          new KeyCloakRsaKeyFetcher()
+              .getPublicKeyFromKeyCloak(
+                  KeyCloakConnectionProvider.SSO_URL, KeyCloakConnectionProvider.SSO_REALM);
+    }
+    return SSO_PUBLIC_KEY;
+  }
+
+  @Override
+  public String verifyToken(String accessToken) throws Exception {
+    try {
+      PublicKey publicKey = getPublicKey();
+      if (publicKey == null) {
+        ProjectLogger.log(
+            "KeyCloakServiceImpl: SSO_PUBLIC_KEY is NULL. Keycloak server may need to be started. Read value from environment variable.",
+            LoggerEnum.INFO);
+        publicKey = toPublicKey(System.getenv(SSOConstant.SSO_PUBLIC_KEY));
```

Comment:
we should remove this code block to read public key from environment variables.
Either this should be read from ENV or to be generated by keycloak rest api.

Reply:
Removed reading from env.